packages/atlas-service/src/util.ts

```diff
@@ -11,16 +11,113 @@ export type IntrospectInfo = { active: boolean };
 
 export type Token = plugin.IdPServerResponse;
 
+function hasExtraneousKeys(obj: any, expectedKeys: string[]) {
+  return (
+    Object.keys(obj).filter((key) => !expectedKeys.includes(key)).length > 0
+  );
+}
+
 export type AIAggregation = {
-  content?: {
+  content: {
     aggregation?: {
-      pipeline?: unknown;
+      pipeline?: string;
     };
   };
 };
 
+export function validateAIAggregationResponse(
+  response: any
+): asserts response is AIAggregation {
+  const { content } = response;
+
+  if (typeof content !== 'object' || content === null) {
+    throw new Error('Unexpected response: expected content to be an object');
+  }
+
+  if (hasExtraneousKeys(content, ['aggregation'])) {
+    throw new Error('Unexpected keys in response: expected aggregation');
+  }
+
+  if (content.aggregation && typeof content.aggregation.pipeline !== 'string') {
+    // Compared to queries where we will always get the `query` field, for
+    // aggregations backend deletes the whole `aggregation` key if pipeline is
+    // empty, so we only validate `pipeline` key if `aggregation` key is present
+    throw new Error(
+      `Unexpected response: expected aggregation to be a string, got ${String(
+        content.aggregation.pipeline
+      )}`
+    );
+  }
+}
+
 export type AIQuery = {
-  content?: {
-    query?: unknown;
+  content: {
+    query: Record<
+      'filter' | 'project' | 'collation' | 'sort' | 'skip' | 'limit',
+      string
+    > & { aggregation?: { pipeline: string } };
   };
 };
+
+export function validateAIQueryResponse(
+  response: any
+): asserts response is AIQuery {
+  const { content } = response ?? {};
+
+  if (typeof content !== 'object' || content === null) {
+    throw new Error('Unexpected response: expected content to be an object');
+  }
+
+  if (hasExtraneousKeys(content, ['query'])) {
+    throw new Error('Unexpected keys in response: expected query');
+  }
+
+  const { query } = content;
+
+  if (typeof query !== 'object' || query === null) {
+    throw new Error('Unexpected response: expected query to be an object');
+  }
+
+  if (
+    hasExtraneousKeys(query, [
+      'filter',
+      'project',
+      'collation',
+      'sort',
+      'skip',
+      'limit',
+      'aggregation',
+    ])
+  ) {
+    throw new Error(
+      'Unexpected keys in response: expected filter, project, collation, sort, skip, limit, aggregation'
+    );
+  }
+
+  for (const field of [
+    'filter',
+    'project',
+    'collation',
+    'sort',
+    'skip',
+    'limit',
+  ]) {
+    if (query[field] && typeof query[field] !== 'string') {
+      throw new Error(
+        `Unexpected response: expected field ${field} to be a string, got ${String(
+          query[field]
+        )}`
+      );
+    }
+  }
+
+  if (query.aggregation && typeof query.aggregation.pipeline !== 'string') {
+    throw new Error(
+      `Unexpected response: expected aggregation pipeline to be a string, got ${String(
+        query.aggregation
+      )}`
```

Comment:
Could we use `util.inspect()` here? It sounds like we’re expecting `query.aggregation` to be an object, so what we’d be getting by default if it just happens to be an object of a different shape is `[object Object]` here (ditto above)

Reply:
Yeah, good idea, haven't thought about that, will update